src/bin/muxer/avformat.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2017-2018, The rav1e contributors. All rights reserved
+//
+// This source code is subject to the terms of the BSD 2 Clause License and
+// the Alliance for Open Media Patent License 1.0. If the BSD 2 Clause License
+// was not distributed with this source code in the LICENSE file, you can
+// obtain it at www.aomedia.org/license/software. If the Alliance for Open
+// Media Patent License 1.0 was not distributed with this source code in the
+// PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+
+#[cfg(feature = "ffmpeg-sys")]
+use ffmpeg_sys::*;
+
+#[allow(unused_imports)]
+use super::Muxer;
+#[allow(unused_imports)]
+use std::ffi::CString;
+#[allow(unused_imports)]
+use std::io;
+#[allow(unused_imports)]
+use std::mem;
+#[allow(unused_imports)]
+use std::ptr;
+
+#[cfg(feature = "ffmpeg-sys")]
+pub struct AvformatMuxer {
+  context: *mut AVFormatContext,
+  duration: i64
+}
+
+#[cfg(feature = "ffmpeg-sys")]
+impl AvformatMuxer {
+  pub fn open(path: &str) -> Box<dyn Muxer> {
+    unsafe {
+      let mut context = ptr::null_mut();
+      let p = CString::new(path).expect("convert to cstring failed");
+      match avformat_alloc_output_context2(
+        &mut context,
+        ptr::null_mut(),
+        ptr::null(),
+        p.as_ptr() as *const i8
+      ) {
+        0 => {
+          let mut io = ptr::null_mut();
+          let ret =
+            avio_open(&mut io, p.as_ptr() as *const i8, AVIO_FLAG_WRITE);
+          if ret < 0 {
+            panic!("open ouput file {} failed", path);
+          }
+          (*context).pb = io;
+          Box::new(AvformatMuxer { context, duration: 0 })
+        }
+        e => panic!("open ouput failed, error = {}", e)
+      }
+    }
+  }
+}
+
+#[cfg(feature = "ffmpeg-sys")]
+impl Muxer for AvformatMuxer {
+  fn write_header(
+    &mut self, width: usize, height: usize, framerate_num: usize,
+    framerate_den: usize
+  ) {
+    unsafe {
+      let stream = avformat_new_stream(self.context, ptr::null_mut());
+      if stream.is_null() {
+        panic!("new stream failed");
+      }
+      let param = (*stream).codecpar;
+      (*param).codec_type = AVMediaType::AVMEDIA_TYPE_VIDEO;
+      (*param).codec_id = AVCodecID::AV_CODEC_ID_AV1;
+      (*param).width = width as i32;
+      (*param).height = height as i32;
+      let ret = avformat_write_header(self.context, ptr::null_mut());
+      if ret < 0 {
+        panic!("write header failed error = {}", ret);
+      }
+      let framerate_den = framerate_den as i32;
+      let framerate_num = framerate_num as i32;
+      let duration = (*stream).time_base.den * framerate_den
+        / ((*stream).time_base.num * framerate_num);
+      self.duration = duration as i64;
+    }
+  }
+
+  fn write_frame(&mut self, pts: u64, data: &[u8]) {
+    unsafe {
+      let mut pkt: AVPacket = mem::zeroed();
```

Comment:
`av_packet_alloc()`?

Reply:
av_write_frame will not take ownership, 
av_init_packet is enough for our case. It's better if we can use uninitialized memory. But I did not find an easy way in rust.

Reply:
'better'?

Reply:
I mean the av_init_packet is enough for this case since av_write_frame will not take the packet ownership.
If we use av_packet_alloc, it introduces unnecessary malloc and we need to call av_packet_free at the end of the function. 
Maybe I missed some important things here. Could you explain why you prefer av_packet_alloc?

Reply:
Mostly just because it avoids uninitialized memory. I have no strong feelings on this.